comms/dht/src/crypt.rs

```diff
@@ -43,17 +49,53 @@ use crate::{
     version::DhtProtocolVersion,
 };
 
+const DOMAIN_SEPARATION_CHALLENGE_LABEL: &str = "com.tari.comms.dht.crypt.challenge";
+const DOMAIN_SEPARATION_KEY_MESSAGE_LABEL: &str = "com.tari.comms.dht.crypt.challenge";
+const DOMAIN_SEPARATION_KEY_SIGNATURE_LABEL: &str = "com.tari.comms.dht.crypt.key_signature";
+
 #[derive(Debug, Clone, Zeroize)]
 #[zeroize(drop)]
 pub struct CipherKey(chacha20::Key);
+pub struct AuthenticatedCipherKey(chacha20poly1305::Key);
+
+// TODO:
+//  1. rename mac_challenge function
+//  2. check if output of generate_ecdh_secret has correct output size
 
 /// Generates a Diffie-Hellman secret `kx.G` as a `chacha20::Key` given secret scalar `k` and public key `P = x.G`.
-pub fn generate_ecdh_secret<PK>(secret_key: &PK::K, public_key: &PK) -> CipherKey
+pub fn generate_ecdh_secret<PK>(secret_key: &PK::K, public_key: &PK) -> [u8; 32]
 where PK: PublicKey + DiffieHellmanSharedSecret<PK = PK> {
     // TODO: PK will still leave the secret in released memory. Implementing Zerioze on RistrettoPublicKey is not
     //       currently possible because (Compressed)RistrettoPoint does not implement it.
     let k = PK::shared_secret(secret_key, public_key);
-    CipherKey(*Key::from_slice(k.as_bytes()))
+    let mut output = [0u8; 32];
+
+    output.copy_from_slice(k.as_bytes());
+    output
+}
+
+pub fn generate_key_message(data: &[u8]) -> CipherKey {
+    // domain separated hash of data (e.g. ecdh shared secret) using hashing API
+    let domain_separated_hash =
+        DomainSeparatedHasher::<Challenge, GenericHashDomain>::new(DOMAIN_SEPARATION_KEY_MESSAGE_LABEL)
+            .chain(data)
+            .finalize()
+            .into_vec();
+
+    // Domain separation uses Challenge = Blake256, thus its output has 32-byte length
+    CipherKey(*Key::from_slice(domain_separated_hash.as_bytes()))
+}
+
+pub fn generate_key_signature_for_authenticated_encryption(data: &[u8]) -> AuthenticatedCipherKey {
+    // domain separated of data (e.g. ecdh shared secret) using hashing API
+    let domain_separated_hash =
+        DomainSeparatedHasher::<Blake256, GenericHashDomain>::new(DOMAIN_SEPARATION_KEY_SIGNATURE_LABEL)
+            .chain(data)
+            .finalize()
+            .into_vec();
+
+    // Domain separation uses Challenge = Blake256, thus its output has 32-byte length
```

Comment:
The use of a specific ECDH shared secret type would mean this function enforces this input type.

Reply:
> It is possible to check and enforce this at compile time, in case later changes subtly break this?

I believe so. Noting that this will not really break sublty and will panic in `Key::from_slice` if not 32-bytes. There are many tests that would signal this breakage, so we can decide if the following is worthwhile.

`digest::Output<D>` returned from `finalize` implements `Into<[u8;32]>` for 32-byte outputs  - this could be used to generate a compile error if the `DomainSeparatedHash<D>` type exposed the inner output type e.g `into_inner(self) -> Output<D>`
```rust
 let domain_separated_hash =
        DomainSeparatedHasher::<Blake256, GenericHashDomain>::new(DOMAIN_SEPARATION_KEY_SIGNATURE_LABEL)
            .chain(data)
            .finalize();
// Will not compile if Into<[u8;32]> is not implemented.
let output_array : [u8;32] = domain_separated_hash.into_inner().into();
```

This requires adding the `into_inner` function or reimplementing (something like) `From<[u8;D::OutputSize::to_usize()]>` on `DomainSeparatedHash`.

Reply:
@sdbondi How generic can that be made? It may be the case that other domain-separated hash function applications require output of a different length.

Reply:
@AaronFeickert That is handled by the [`generic_array` crate](http://fizyk20.github.io/generic-array/generic_array/). Here, we are enforcing in this function (or else compile fails) that the hash output is correctly convertible to a 32-byte array representation.

Adding trait bounds are not necessary here as we use concrete types, but the trait bounds would be `where digest::Output<D>: Into<[u8;32]>`

Reply:
So as the tari_crypto domain hasher code currently stands, you'd need to disambiguate between the functions on the hasher type and the `impl Digest` functions with the same name.
e.g.
```rust
     let hasher = comms_dht_hash_domain_key_signature().chain(data);
    let domain_separated_hash =<DomainSeparatedHasher<CommsChallenge, DHTCommsHashDomain> as Digest>::finalize(hasher);
    let hash: [u8; 32] = domain_separated_hash.into();
```

Since we know that we're using Blake256 and breakage will be caught in tests - I think we can leave this as is